pkg/envinfo/storage.go

```diff
@@ -64,11 +72,45 @@ func (ei *EnvInfo) CreateStorage(storage localConfigProvider.LocalStorage) error
 				},
 			},
 		},
-	}, storage.Path)
-
+	}
+    //2. Iterate over components in devfile
+	components, err := ei.devfileObj.Data.GetComponents(common.DevfileOptions{})
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get devfile components while creating storage %w", err)
 	}
+	for _, component := range components {
+		// if component has container and if container to attach is not provided or it is provided and name component matched
+		// only then check if storage is not already mounted.
+		if component.Container != nil && (storage.Container == "" || storage.Container != "" && component.Volume == nil && storage.Container == component.Name) {
+			for _, vm := range component.Container.VolumeMounts {
+				if vm.Path == storage.Path {
+					var err = fmt.Errorf("another volume, %s, is mounted to the same path: %s, on the container: %s", vm.Name, storage.Path, component.Name)
+					pathErrorContainers = append(pathErrorContainers, err.Error())
+				}
+			}
+			if len(pathErrorContainers) == 0 {
+				component.Container.VolumeMounts = append(component.Container.VolumeMounts, volumeMount)
+			}
+		} else if component.Volume != nil && component.Name == storage.Name {
+			volumeExists = true
+		}
+	}
+
+	//if volume does not already exist, add it. Otherwise update it
+	if volumeExists {
+		ei.devfileObj.Data.UpdateComponent(volumeComponent)
+	} else {
+		err = ei.devfileObj.Data.AddComponents([]devfilev1.Component{volumeComponent})
```

Comment:
here as we are checking `pathErrorContainers` below this, we might be adding duplicate volume mounts?

Reply:
Moved to above the loop. I will push the updated commit shortly.

I am not sure though how we will add duplicate volume mounts. We only add volume mount, if the container in question does not already have it, and similarly, we add the global volume component only if it has not already been added. Otherwise, we update the existing volume component